### ISSUE TITLE
OIDC redirects from team site to personal page after login

### ISSUE THREAD

**Issue by helmut72** [gristlabs/grist-core#740]
Thanks a lot for OIDC! Works great in my limited tests yesterday. Public Sharing, Team site ... all works, fantastic! Only found two minor issues.

One of them...

After creating a team site...

```
curl -s -X POST -H "Content-Type: application/json" -d '{"name": "Outside Sensors", "domain": "outside"}' -H "Authorization: Bearer <my api key>" https://docs.grist.example.com/api/orgs
```

... and opening `https://outside.grist.example.com`, I will be redirect to Authelia (or Keycloak). After login, Authelia redirects to my personal page `https://docs.grist.example.com`. SAML redirects back to the team site.

This is very minor and I'm not sure if this works at all with OIDC. Only a feedback. If this can't work with OIDC, just close my issue.

Thanks.

**Comment by dsagal** [comment]
This should be fixable. The method [getLoginRedirectUrl()](https://github.com/gristlabs/grist-core/blob/5f0c3bd8545256b4d3310ace197aace284ce8633/app/server/lib/OIDCConfig.ts#L118) takes a second parameter (`targetUrl: URL`), which is the final destination URL. We can relay that to the end of the flow by either including it into the `state` parameter (along with the random value) or into the session (which seems simpler).

Let me cc @fflorent, since some of the points below are relevant to OIDC whether or not we deal with redirects.

Regarding the `state` parameter, it's unclear if it's actually important to security, since pkce (code_challenge / code_verifier) address the same need and more. From what I could find online though, `state` is still recommended to include (some of that debate is visible here: https://security.stackexchange.com/a/215027). Assuming we keep `state`, it seems to me that the code [here](https://github.com/gristlabs/grist-core/blob/5f0c3bd8545256b4d3310ace197aace284ce8633/app/server/lib/OIDCConfig.ts#L90) isn't quite right: `state` should be stored in the session, to be compared inside the callback to the state passed back from the authorization server.

In other words, when saving values in the session, I think it would make sense to save this:
```
    mreq.session.oidc = {
      codeVerifier,
      state,
      targetUrl,
    };
```
Then use the `state` from here when calling `this._client.callback()`, and use `targetUrl` from here for the final redirect.

Something I hadn't thought of earlier, but I think should be added, is clearing `mreq.session.oidc` as soon as those values are used, and even in case of error. I saw this recommendation somewhere where I can't find at the moment, but it makes sense to me, since these are intended as single-use values.

It's important that attackers can't set `targetUrl` to anything they want -- setting it to a phishing site would be bad. I think the value of `targetUrl` passed in to `getLoginRedirectUrl()` is already safe (at least several code paths I looked at seem to take care of it); but it would be a good idea to add a test case that it's not easy to get Grist to redirect one to a random site after an OIDC login.
